### DISCUSSION TITLE
MAUI Wear OS/Watch OS Support

### DISCUSSION THREAD

**Discussion by DevronB** [dotnet/maui#1144] [Q&A]
Hi all, I've developed a wearable app on Samsungs Tizen platform and have been developing using the Tizen Wearable Web frameworks. Samsung have announced that they will be moving from Tizen Wearable to Android - Wear OS. I'm not sure where this leaves me as a Tizen developer as I understand Wear OS does not support WebView. I'm currently evaluating my options for development on Wear OS. I know there's Wear OS Xamarin... but my questions is:  Would the MAUI framework ever support Wear OS/Watch OS, I understand that the "Forms" most likely won't/can't but could I develop and app using dotnet 6 and target Wear OS even if I had to write the UI using native control/libs myself over XAML. If I could then use MAUI for the mobile parts of the app this would be ideal.... and maybe target Andriod and iOS. My preference is to develop with dotnet C#.  As you can probably guess from my question I'm pretty new to Wear OS, Android, MAUI and Xamarin, but not so much dotnet.  

I don't fully understand how these system are targeted and their compatibility so any advice/comments would be appreciated.

**Comment by DevronB** [comment]
As there is no response on this I will ask the question a different way.

As MAUI is the progression of Xamarin and Xamarin currently supports Wear OS/Watch OS, will MAUI support Wear OS/Watch OS for dotnet 6 development and code sharing?

**Reply by EmilAlipiev** [reply]
Code sharing you can do but UI is not easy to unify. thats why Xamarin.Forms never supported it. Not many devs are doing watch development also because it doesnt bring any economical benefit. You want also find many (even i have never seen any) company looking for a watch developer. Thats why I dont think for hobby development Microsoft will bring Crossplatform for watches. It will be nice to have Xamarin.Native like posibility with MAUI.

**Reply by nitinjs** [reply]
Blazor hybrid MAUI is supporting Tizen OS why not Wear OS?

**Comment by DevronB** [comment]
@Clancey, I've been watching some of your videos on Maui and Comet and thought you may be able to comment on this?

I'm looking to write a Android phone/Wear OS app and would like mostly the same code base. Ideally I would use MAUI with MVU/Comet and maybe some complier directives to switch UI based on Android/Wear OS. 

I can get a basic Maui Android app running on Wear OS Emulator so I assume the core should work and be supported.
I could import Andriodx.Wear nuget packages, but would the Wear specific controls need to be ported to Maui somehow? Is there any plans for WearOS/WatchOS to be nodes of a Maui project?

I guess I'm looking for some direction on where I should start with a Android/Phone/Wear app while future proofing my efforts. I'd be happy to run with Maui from what I have seen but I'm not sure about the Wear OS side of things.

I've been doing a lot of research and am still a little lost as to where I should start. Any advice would be appreciated.

**Comment by Zumpel96** [comment]
Clarification regarding this is highly appreciated. FINALLY the smartwatch manufacturers are moving towards a unified operating system and this would make the development so much easier, if it would be possible to use the Wear OS platform right next to the Android or even iOS platform.

**Comment by BobbyCannon** [comment]
I'm very interested in this also.

**Comment by NabodevMats** [comment]
I also need answer on this.

Edit: just found this which doesn't state Wear OS at this time:
[https://docs.microsoft.com/en-us/dotnet/maui/supported-platforms](https://docs.microsoft.com/en-us/dotnet/maui/supported-platforms)

**Comment by hez2010** [comment]
See progress here: https://github.com/dotnet/maui/pull/2360

**Reply by BobbyCannon** [reply]
Maybe some confusion though? Or maybe I'm confused. Tizen support is interesting but I'm interested in Wear OS support. Did I miss something?

**Reply by hez2010** [reply]
Never mind, I thought you were asking the Tizen backend

**Comment by mcblacksea** [comment]
The same here.
We'd be glad to see the Watch OS and Wear OS development options with .NET MAUI release.

**Comment by Zumpel96** [comment]
It is a little bit frustrating to see no response. Tizen will be supported, as it seems and the funny thing: **Samsung moves away from Tizen!!!** Their newest generation of smartwatches (the Watch4) moved over to Wear OS!

e:// Yes, I know, that Tizen is more than just for smartwatches. But them moving away from their own OS is something, which should be considered.

**Reply by EmilAlipiev** [reply]
Tizen isnt only watch. I think that they continue more with mobile and TV in priority. Tizen isnt using Android SDK also, it has its own SDK and they wrap around their own SDK. so it is not developed by Xamarin team but Samsung Team.

**Comment by BobbyCannon** [comment]
Any updates on this?

**Reply by DevronB** [reply]
I did some trials and built a MAUI (and Comet) project and deployed to Wear OS. Startup time was way too slow on wearables. I'm not sure about the Watch OS support either. I ended up building an net6-android project and stuck with the native wrappers which seems to work fine... but ideally we could have a MAUI project that targets Android, iOS, WearOS and WatchOS. Then we could just use conditional code for the different systems where the diverge. I have no idea weather this is on the MAUI roadmap as per this thread.

**Reply by DavidH102** [reply]
I don't think it's on the roadmap anywhere that I can see?

**Reply by EmilAlipiev** [reply]
Sure, you cannot build WearOS using XF also and Cross platform not supported. That will be nice but I dont think that it is coming soon. So only Native development possible, like Xamarin.Native

**Comment by noumanqaiser** [comment]
A Clear dotnet Roadmap for wearable devices is a n absolute necessity, is MAUI going to support WearOS anytime soon?

**Reply by JohnHDev** [reply]
Maui is a progression of Xamarin.Forms, and XF has never supported wearables.
What we really want to see is native C# support for wearables and widgets too.

**Reply by EmilAlipiev** [reply]
@JohnHDev so for native still we need to keep using Xamarin.Native? If they keep it up to date parallel with Maui it is still ok.

**Reply by JohnHDev** [reply]
Not Xamarin.Native, but .net6 for iOS and for Android.

**Comment by beeradmoore** [comment]
So update on this via [Twitter](https://twitter.com/davidortinau/status/1519303524726849536). WatchOS (and I assume WearOS but not 100%) is not coming with .NET 6 and currently isn't planned for .NET 7.

Via @davidortinau ,

> we don’t have watchOS support in .NET 6 and no plans to add it in .NET 7. The original .NET 6 issue tracking Apple SDK plans stated this and since then it’s not on any roadmap. We will consider it in the future. High cost, low usage.

**Reply by chucker** [reply]
(FWIW, .NET 6 not including watchOS support was originally announced at https://github.com/xamarin/xamarin-macios/issues/8955. Also, it not supporting MAUI is *on top of* that; it first needs to be supported at all in .NET before MAUI can even be a thing.)

**Reply by beeradmoore** [reply]
That makes sense (first .NET 6 and then MAUI).

Just not a great thing to wake up to, that my apps suddenly have an end of life. 

From what I see Xamarin.iOS and Xamarin.Android will continue to be supported for 2 years at the end of their final release. I am unsure yet when that is but they both are moving to net6.0-ios and net6.0-android.  With the new greatness which is net6.0-* it means no longer supporting watchOS/WearOS, while also being able to continue support Tizen based watch devices (thank you Samsung I guess).

Only way for me to continue to support these apps is to port them to Kotlin/Java and ObjC/swift. That most probably won't happen. So next best thing is to hope to get it into .NET 7 or .NET 8. Otherwise I am SoL.

**Reply by vhugogarcia** [reply]
It won't be supported by MAUI, however, that does not mean you cannot develop for watchOS or Android Wear, you can still develop without problems independently and share all the business logic between platforms natively and use VS for Mac or Windows to develop to those platforms. 

MAUI won't support it because the fact the UI elements are completely different, that is all.

**Reply by beeradmoore** [reply]
Can you please make a demo of a WatchOS or wearOS app working with a MAUI app?

.NET 6 does not support building WatchOS/wearos. So you either need to have the mono project equivalent inside your MAUI solution which goes end of life 2 years after latest release. I was trying this out yesterday but couldn't get just the net6.0-ios project to be the only one referencing the watch app.

or 

go down the path I am now currently experimenting which is fully native WatchOS/wearOS app compiled with native tools and bundled within the compiled MAUI bundle.

If you have an example of either of these working it would be greatly appreciated?

**Reply by chucker** [reply]
> It won't be supported by MAUI, however, that does not mean you cannot develop for watchOS or Android Wear

MAUI aside, there simply is no .NET 6 (or beyond) tooling for watchOS at this point. You’d have to use the Xamarin.watchOS tooling, which I imagine will work worse and worse over time (and is *already* quite hit and miss).

**Reply by beeradmoore** [reply]
I've started experimenting with that. So far for WearOS it appears we can continue to use Xamarin.Android tooling, and in theory can use native WearOS apps for that platform. Playing around more with watchOS next.

If anyone wants to give it a try feel free to fork and send a PR, or DM me directly on Twitter under my same handle if you wanted to type talk some more without spamming this thread.

http://github.com/beeradmoore/MAUI-Watch-n-Wear

**Comment by ToolmakerSteve** [comment]
To clarify: This **should not be considered a Maui topic**. Its really about the migration of **Xamarin native** to **.Net 6, 7, etc**, and what support **.Net** will have for `watchOS`.

IMHO, it would not be sane to expect to develop a cross-platform UI, and have it run well on watchOS. Rather, one should use whatever APIs watchOS provides. That makes this a `.Net` futures question, not a `Maui` question.

**Comment by akhanalcs** [comment]
Hopefully MAUI adds this in the future. :)

**Reply by acaliaro** [reply]
https://www.youtube.com/watch?v=OoNyWM_oGlU

**Reply by NabodevMats** [reply]
<img width="250" alt="image" src="https://github.com/dotnet/maui/assets/20288541/4eb4dc47-f993-4544-9b09-84f65950a3ff">
?

**Comment by DamianSuess** [comment]
Im crossing fingers for a future integration of Android Wear. We're upgrading our product base from Xamarin.Forms to MAUI and this is a hit to us as well. First, it was Samsung hit was dropping Tizen on their watches in favor for Android.

Really would like to keep a singular codebase and not have to hop languages to support the watches.

**Reply by acaliaro** [reply]
Something can be done https://github.com/acaliaro/MauiAppWearOsPicking

**Reply by DamianSuess** [reply]
Alessandro, seeing your reply just made my day. Not only from your repository reply but also it coming from you, a name I recall seeing on the original Xamarin forums years back. 👍  Thank you

**Reply by acaliaro** [reply]
old but good

**Reply by knocte** [reply]
> ... Samsung hit was dropping Tizen on their watches in favor for Android.

Can you give a link about this? I've found this: https://soyacincau.com/2021/07/11/samsung-galaxy-gear-android-to-tizen-os-upgrade/ but it sounds as speculation, given sentences are in the future tense such as "Not just that, the Galaxy Watch 4 is also said to feature the Google Play store for the first time".

**Comment by arvindrajachourasiya** [comment]
Hi, is there any change, can we develop watch application using .Net MAUI ?

**Comment by LennoxP90** [comment]
i would like to see this as a reality as well

**Comment by zandiarash** [comment]
Newly Thank to **Dotnet 8 ``Preview``** and **Visual Studio Version 17.8.0 ``Preview 5.0``** seems something is happening.
My hello world project works like a charm even I did not customized **AndroidManifest.xml** file
Please Look at my project :

![image](https://github.com/dotnet/maui/assets/10775073/a70fe71f-e750-4103-bec5-2d22d9db6ac1)


Tip : you should ``**not**`` customize **AndroidManifest.xml** and it should be empty like this :
![image](https://github.com/dotnet/maui/assets/10775073/b8ba40ca-4b7f-423c-b355-02d3158cb35b)


It should ``**not**`` have the ``<Uses-SDK`` line
![image](https://github.com/dotnet/maui/assets/10775073/93a00857-6fe7-4196-8313-c79bab52cf2e)

**Comment by justintemplar** [comment]
Vote for Apple Watch OS to please get some attention.

**Reply by zandiarash** [reply]
Hi, where should I vote ?

**Comment by axa88** [comment]
> Newly Thank to **Dotnet 8 `Preview`** and **Visual Studio Version 17.8.0 `Preview 5.0`** seems something is happening. My hello world project works like a charm even I did not customized **AndroidManifest.xml** file Please Look at my project 


Fwiw I could run a default VS temple project on my Galaxy Watch 4 ever since .NET 7
Surely things must have improved, but merely running a default template isn't one

**Reply by beeradmoore** [reply]
That's because it deploys it as an Android app to run on WearOS through some magic. AFAIK you should not do this for WearOS apps as the Android type and not using WearOS optimised components. 

This goes back to .NET 6 I think. In [mid 2022](https://twitter.com/beeradmoore/status/1529276251030298624) I had selected the wrong connected device to deploy to and didn't realise. Saw a circle purple flash off on the corner of my desk as my old TicWatch I had plugged in to charge had become my default deploy target 😂

![image](https://github.com/dotnet/maui/assets/904737/3a14899d-05bf-4c7f-a68d-e4454852c72f)

**Reply by axa88** [reply]
I've been doing this now for over a year as it suits my limited use fine. Of course it would be a challenge to publish and distribute without being able to specifically target WearOS but I don't care to do so.

Fwiw there's no "magic" nor optimization, only wear specific UI components and a number of minor additions like ambient mode and couple of subtle limitations due to battery concerns. Afaik.

Hope it gets official support someday

**Comment by nowlights** [comment]
I have an application developed in Xamarin that has an extension for the Apple Watch OS, and my question is, if I update my Xamarin app to .NET MAUI, will I lose this extension?

Will my Apple Watch app be removed from the library?
Will I no longer be able to work on improvements and fixes for the Apple Watch app?

**Reply by beeradmoore** [reply]
If your WatchOS app is built with Xamarin, yes. There is no .NET support for WatchOS apps.

Currently you need to re-write your WatchOS app in swift/obj-c and then bundle it with your MAUI app.

**Reply by nowlights** [reply]
So can I continue working on my Android/IOS with MAUI and on WatchOS using xcode with Swift? Is it perfectly possible?

**Reply by beeradmoore** [reply]
Correct, perfectly possible.

 I don't remember all the changes, but you need to link it into your csproj in a different way.

**Reply by nowlights** [reply]
but that's the problem, how can i do that rsrsrs

**Reply by beeradmoore** [reply]
As far as I know there is no official documentation, 

big thread on it here, https://github.com/xamarin/xamarin-macios/issues/10070
but most importantly working (I assume, I am not using it at the moment) samples here, https://github.com/vouksh/MauiWithWatchApps/

As an extra there is some recent discussion on adding watchOS back in here, https://github.com/xamarin/xamarin-macios/issues/20317

**Comment by takrilioy** [comment]
I have two different projects, one for WearOS (and possible WatchOs) and one for table/phone apps I've managed to develop and publish Android App, Android WearOs App and iOS Phone iPad App. I don't have Mac, so I've used Github Actions to compile iOS versions. By adding property <key>WKWatchKitApp</key><true/>  to Info.plist file in WearOs project the compiler give an error ".NET MAUI Assets.xcassets : actool warning : A 1024x1024 App Store icon is required for watchOS apps". Does anyone know how to tackle this? I tried to set 1024x1024 icon as AppIcon with base size 1024,1024 but it didn't work? I know that official support is not available, but as WearOs was possible (with quite a lot of tweaking for round devices) I suspect that it could be possible also somehow f.ex generating assets by hand.
